File: SignalServiceKit/src/Security/OWSFingerprint.m
//
// Copyright 2017 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "OWSFingerprint.h"
#import "NSData+keyVersionByte.h"
#import "OWSError.h"
#import <CommonCrypto/CommonDigest.h>
#import <SignalCoreKit/NSData+OWS.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>
#import <UIKit/UIImage.h>

NS_ASSUME_NONNULL_BEGIN

static uint32_t const OWSFingerprintHashingVersion = 0;
static uint32_t const OWSFingerprintPreUUIDScannableFormatVersion = 1;
static uint32_t const OWSFingerprintScannableFormatVersion = 2;
static uint32_t const OWSFingerprintDefaultHashIterations = 5200;

@interface OWSFingerprint ()

@property (nonatomic, readonly) NSUInteger hashIterations;
@property (nonatomic, readonly) NSString *text;
@property (nonatomic, readonly) NSData *myFingerprintData;
@property (nonatomic, readonly) NSData *theirFingerprintData;
@property (nonatomic, readonly) NSString *theirName;

@end

#pragma mark -

@implementation OWSFingerprint

- (instancetype)initWithMyStableAddress:(SignalServiceAddress *)myStableAddress
                          myIdentityKey:(NSData *)myIdentityKeyWithoutKeyType
                     theirStableAddress:(nonnull SignalServiceAddress *)theirStableAddress
                       theirIdentityKey:(NSData *)theirIdentityKeyWithoutKeyType
                              theirName:(NSString *)theirName
                         hashIterations:(uint32_t)hashIterations
{
    OWSAssertDebug(theirIdentityKeyWithoutKeyType.length == 32);
    OWSAssertDebug(myIdentityKeyWithoutKeyType.length == 32);
    OWSAssertDebug(myStableAddress.isValid);
    OWSAssertDebug(theirStableAddress.isValid);

    self = [super init];
    if (!self) {
        return self;
    }

    _myStableAddress = myStableAddress;
    _myIdentityKey = [myIdentityKeyWithoutKeyType prependKeyType];
    _theirStableAddress = theirStableAddress;
    _theirIdentityKey = [theirIdentityKeyWithoutKeyType prependKeyType];
    _theirName = theirName;
    _hashIterations = hashIterations;

    NSData *myStableAddressData = [self stableDataForAddress:_myStableAddress];
    _myFingerprintData = [self dataForStableAddress:myStableAddressData publicKey:_myIdentityKey];

    NSData *theirStableAddressData = [self stableDataForAddress:_theirStableAddress];
    _theirFingerprintData = [self dataForStableAddress:theirStableAddressData publicKey:_theirIdentityKey];

    return self;
}

+ (instancetype)fingerprintWithMyStableAddress:(SignalServiceAddress *)myStableAddress
                                 myIdentityKey:(NSData *)myIdentityKeyWithoutKeyType
                            theirStableAddress:(nonnull SignalServiceAddress *)theirStableAddress
                              theirIdentityKey:(NSData *)theirIdentityKeyWithoutKeyType
                                     theirName:(NSString *)theirName
                                hashIterations:(uint32_t)hashIterations
{
    return [[self alloc] initWithMyStableAddress:myStableAddress
                                   myIdentityKey:myIdentityKeyWithoutKeyType
                              theirStableAddress:theirStableAddress
                                theirIdentityKey:theirIdentityKeyWithoutKeyType
                                       theirName:theirName
                                  hashIterations:hashIterations];
}

+ (instancetype)fingerprintWithMyStableAddress:(SignalServiceAddress *)myStableAddress
                                 myIdentityKey:(NSData *)myIdentityKeyWithoutKeyType
                            theirStableAddress:(nonnull SignalServiceAddress *)theirStableAddress
                              theirIdentityKey:(NSData *)theirIdentityKeyWithoutKeyType
                                     theirName:(NSString *)theirName
{
    return [[self alloc] initWithMyStableAddress:myStableAddress
                                   myIdentityKey:myIdentityKeyWithoutKeyType
                              theirStableAddress:theirStableAddress
                                theirIdentityKey:theirIdentityKeyWithoutKeyType
                                       theirName:theirName
                                  hashIterations:OWSFingerprintDefaultHashIterations];
}

- (uint32_t)scannableFingerprintVersion
{
    if (!RemoteConfig.uuidSafetyNumbers) {
        return OWSFingerprintPreUUIDScannableFormatVersion;
    }

    return OWSFingerprintScannableFormatVersion;
}

- (NSData *)stableDataForAddress:(SignalServiceAddress *)address
{
    // For now, leave safety number based on phone number unless the feature flag is enabled.
    // This prevents mismatch from occurring against old apps until we formally roll out the feature.
    if (RemoteConfig.uuidSafetyNumbers && address.uuid != nil) {
        // TODO UUID: Right now, uuid is nullable, but safety numbers require us to always have
        // the UUID for a user. This will need to be updated once we change this field to nonnull.
        NSUUID *uuid = address.uuid;
        OWSAssert(uuid);
        uuid_t bytes;
        [uuid getUUIDBytes:bytes];
        return [[NSData alloc] initWithBytes:bytes length:16];
    } else {
        NSString *phoneNumber = address.phoneNumber;
        OWSAssert(phoneNumber);
        OWSAssert(phoneNumber.length > 0);
        return [phoneNumber dataUsingEncoding:NSUTF8StringEncoding];
    }
}

- (BOOL)matchesLogicalFingerprintsData:(NSData *)data error:(NSError **)error
{
    OWSAssertDebug(data.length > 0);
    OWSAssertDebug(error);

    *error = nil;
    FingerprintProtoLogicalFingerprints *_Nullable logicalFingerprints;
    logicalFingerprints = [[FingerprintProtoLogicalFingerprints alloc] initWithSerializedData:data error:error];
    if (!logicalFingerprints || *error) {
        OWSFailDebug(@"fingerprint failure: %@", *error);

        NSString *description = OWSLocalizedString(@"PRIVACY_VERIFICATION_FAILURE_INVALID_QRCODE", @"alert body");
        *error = [OWSError withError:OWSErrorCodePrivacyVerificationFailure description:description isRetryable:NO];
        return NO;
    }

    if (logicalFingerprints.version < self.scannableFingerprintVersion) {
        OWSLogWarn(@"Verification failed. They're running an old version.");
        NSString *description
            = OWSLocalizedString(@"PRIVACY_VERIFICATION_FAILED_WITH_OLD_REMOTE_VERSION", @"alert body");
        *error = [OWSError withError:OWSErrorCodePrivacyVerificationFailure description:description isRetryable:NO];
        return NO;
    }

    if (logicalFingerprints.version > self.scannableFingerprintVersion) {
        OWSLogWarn(@"Verification failed. We're running an old version.");
        NSString *description = OWSLocalizedString(@"PRIVACY_VERIFICATION_FAILED_WITH_OLD_LOCAL_VERSION", @"alert body");
        *error = [OWSError withError:OWSErrorCodePrivacyVerificationFailure description:description isRetryable:NO];
        return NO;
    }

    // Their local is *our* remote.
    FingerprintProtoLogicalFingerprint *localFingerprint = logicalFingerprints.remoteFingerprint;
    FingerprintProtoLogicalFingerprint *remoteFingerprint = logicalFingerprints.localFingerprint;

    if (![remoteFingerprint.identityData isEqual:[self scannableData:self.theirFingerprintData]]) {
        OWSLogWarn(@"Verification failed. We have the wrong fingerprint for them");
        NSString *descriptionFormat = OWSLocalizedString(@"PRIVACY_VERIFICATION_FAILED_I_HAVE_WRONG_KEY_FOR_THEM",
            @"Alert body when verifying with {{contact name}}");
        NSString *description = [NSString stringWithFormat:descriptionFormat, self.theirName];
        *error = [OWSError withError:OWSErrorCodePrivacyVerificationFailure description:description isRetryable:NO];
        return NO;
    }

    if (![localFingerprint.identityData isEqual:[self scannableData:self.myFingerprintData]]) {
        OWSLogWarn(@"Verification failed. They have the wrong fingerprint for us");
        NSString *descriptionFormat = OWSLocalizedString(@"PRIVACY_VERIFICATION_FAILED_THEY_HAVE_WRONG_KEY_FOR_ME",
            @"Alert body when verifying with {{contact name}}");
        NSString *description = [NSString stringWithFormat:descriptionFormat, self.theirName];
        *error = [OWSError withError:OWSErrorCodePrivacyVerificationFailure description:description isRetryable:NO];
        return NO;
    }

    OWSLogWarn(@"Verification Succeeded.");
    return YES;
}

- (NSString *)text
{
    NSString *myDisplayString = [self stringForFingerprintData:self.myFingerprintData];
    NSString *theirDisplayString = [self stringForFingerprintData:self.theirFingerprintData];

    if ([theirDisplayString compare:myDisplayString] == NSOrderedAscending) {
        return [NSString stringWithFormat:@"%@%@", theirDisplayString, myDisplayString];
    } else {
        return [NSString stringWithFormat:@"%@%@", myDisplayString, theirDisplayString];
    }
}

/**
 * Formats numeric fingerprint, 3 lines in groups of 5 digits.
 */
- (NSString *)displayableText
{
    NSString *input = self.text;

    NSMutableArray<NSString *> *lines = [NSMutableArray new];

    NSUInteger lineLength = self.text.length / 3;
    for (uint i = 0; i < 3; i++) {
        NSString *line = [input substringWithRange:NSMakeRange(i * lineLength, lineLength)];

        NSMutableArray<NSString *> *chunks = [NSMutableArray new];
        for (uint i = 0; i < line.length / 5; i++) {
            NSString *nextChunk = [line substringWithRange:NSMakeRange(i * 5, 5)];
            [chunks addObject:nextChunk];
        }
        [lines addObject:[chunks componentsJoinedByString:@" "]];
    }

    return [lines componentsJoinedByString:@"\n"];
}


- (NSData *)dataFromShort:(uint32_t)aShort
{
    uint8_t bytes[] = {
        ((uint8_t)(aShort & 0xFF00) >> 8),
        (uint8_t)(aShort & 0x00FF)
    };

    return [NSData dataWithBytes:bytes length:2];
}

/**
 * An identifier for a mutable public key, belonging to an immutable identifier (stableId).
 *
 * This method is intended to be somewhat expensive to produce in order to be brute force adverse.
 *
 * @param stableAddressData
 *      Immutable global identifier e.g. Signal Identifier, an e164 formatted phone number encoded as UTF-8 data
 * @param publicKey
 *      The current public key for <stableAddress>
 * @return
 *      All-number textual representation
 */
- (NSData *)dataForStableAddress:(NSData *)stableAddressData publicKey:(NSData *)publicKey
{
    OWSAssertDebug(stableAddressData);
    OWSAssertDebug(publicKey);

    NSData *versionData = [self dataFromShort:OWSFingerprintHashingVersion];
    NSMutableData *hash = [versionData mutableCopy];
    [hash appendData:publicKey];
    [hash appendData:stableAddressData];

    NSMutableData *_Nullable digestData = [[NSMutableData alloc] initWithLength:CC_SHA512_DIGEST_LENGTH];
    if (!digestData) {
        @throw [NSException exceptionWithName:NSGenericException reason:@"Couldn't allocate buffer." userInfo:nil];
    }
    for (int i = 0; i < self.hashIterations; i++) {
        [hash appendData:publicKey];

        if (hash.length >= UINT32_MAX) {
            @throw [NSException exceptionWithName:@"Oversize Data" reason:@"Oversize hash." userInfo:nil];
        }

        CC_SHA512(hash.bytes, (uint32_t)hash.length, digestData.mutableBytes);
        // TODO get rid of this loop-allocation
        hash = [digestData mutableCopy];
    }

    return [hash copy];
}


- (NSString *)stringForFingerprintData:(NSData *)data
{
    OWSAssertDebug(data);

    return [NSString stringWithFormat:@"%@%@%@%@%@%@",
                     [self encodedChunkFromData:data offset:0],
                     [self encodedChunkFromData:data offset:5],
                     [self encodedChunkFromData:data offset:10],
                     [self encodedChunkFromData:data offset:15],
                     [self encodedChunkFromData:data offset:20],
                     [self encodedChunkFromData:data offset:25]];
}

- (NSString *)encodedChunkFromData:(NSData *)data offset:(uint)offset
{
    OWSAssertDebug(data);

    uint8_t fiveBytes[5];
    [data getBytes:fiveBytes range:NSMakeRange(offset, 5)];

    int chunk = [self uint64From5Bytes:fiveBytes] % 100000;
    return [NSString stringWithFormat:@"%05d", chunk];
}

- (int64_t)uint64From5Bytes:(uint8_t[])bytes
{
    int64_t result = ((bytes[0] & 0xffLL) << 32) |
           ((bytes[1] & 0xffLL) << 24) |
           ((bytes[2] & 0xffLL) << 16) |
           ((bytes[3] & 0xffLL) <<  8) |
           ((bytes[4] & 0xffLL));

    return result;
}

- (NSData *)scannableData:(NSData *)data
{
    return [data subdataWithRange:NSMakeRange(0, 32)];
}

- (nullable UIImage *)image
{
    FingerprintProtoLogicalFingerprintBuilder *remoteFingerprintBuilder =
        [FingerprintProtoLogicalFingerprint builderWithIdentityData:[self scannableData:self.theirFingerprintData]];
    NSError *error;
    FingerprintProtoLogicalFingerprint *_Nullable remoteFingerprint =
        [remoteFingerprintBuilder buildAndReturnError:&error];
    if (!remoteFingerprint || error) {
        OWSFailDebug(@"could not build proto: %@", error);
        return nil;
    }

    FingerprintProtoLogicalFingerprintBuilder *localFingerprintBuilder =
        [FingerprintProtoLogicalFingerprint builderWithIdentityData:[self scannableData:self.myFingerprintData]];
    FingerprintProtoLogicalFingerprint *_Nullable localFingerprint =
        [localFingerprintBuilder buildAndReturnError:&error];
    if (!localFingerprint || error) {
        OWSFailDebug(@"could not build proto: %@", error);
        return nil;
    }

    FingerprintProtoLogicalFingerprintsBuilder *logicalFingerprintsBuilder =
        [FingerprintProtoLogicalFingerprints builderWithVersion:self.scannableFingerprintVersion
                                               localFingerprint:localFingerprint
                                              remoteFingerprint:remoteFingerprint];

    // Build ByteMode QR (Latin-1 encodable data)
    NSData *_Nullable fingerprintData = [logicalFingerprintsBuilder buildSerializedDataAndReturnError:&error];
    if (!fingerprintData || error) {
        OWSFailDebug(@"could not serialize proto: %@", error);
        return nil;
    }

    OWSLogDebug(@"Building fingerprint with data: %@", fingerprintData);

    CIFilter *filter = [CIFilter filterWithName:@"CIQRCodeGenerator"];
    [filter setDefaults];
    [filter setValue:fingerprintData forKey:@"inputMessage"];

    CIImage *ciImage = [filter outputImage];
    if (!ciImage) {
        OWSLogError(@"Failed to create QR image from fingerprint text: %@", self.text);
        return nil;
    }

    // UIImages backed by a CIImage won't render without antialiasing, so we convert the backign image to a CGImage,
    // which can be scaled crisply.
    CIContext *context = [CIContext contextWithOptions:nil];
    CGImageRef cgImage = [context createCGImage:ciImage fromRect:ciImage.extent];
    UIImage *qrImage = [UIImage imageWithCGImage:cgImage];
    CGImageRelease(cgImage);

    return qrImage;
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/Security/OWSFingerprint.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import CommonCrypto
import LibSignalClient

public class OWSFingerprint {
    public let myAci: Aci
    public let theirAci: Aci

    public let myAciIdentityKey: Data
    public let theirAciIdentityKey: Data

    private let myFingerprintData: Data
    private let theirFingerprintData: Data

    private let hashIterations: UInt32
    public let theirName: String

    /**
     * Formats numeric fingerprint, 3 lines in groups of 5 digits.
     */
    public var displayableText: String {
        return generateDisplayableText()
    }

    public var image: UIImage? {
        return generateImage()
    }

    public init(
        myAci: Aci,
        theirAci: Aci,
        myAciIdentityKey: Data,
        theirAciIdentityKey: Data,
        theirName: String,
        hashIterations: UInt32 = Constants.defaultHashIterations
    ) {
        self.myAci = myAci
        self.theirAci = theirAci
        let myAciIdentityKey = myAciIdentityKey.prependKeyType()
        self.myAciIdentityKey = myAciIdentityKey
        let theirAciIdentityKey = theirAciIdentityKey.prependKeyType()
        self.theirAciIdentityKey = theirAciIdentityKey
        self.hashIterations = hashIterations
        self.theirName = theirName

        let (myStableSourceData, theirStableSourceData) = Self.stableData(myAci: myAci, theirAci: theirAci)
        self.myFingerprintData = Self.dataForStableAddress(
            myStableSourceData,
            publicKey: myAciIdentityKey,
            hashIterations: hashIterations
        )
        self.theirFingerprintData = Self.dataForStableAddress(
            theirStableSourceData,
            publicKey: theirAciIdentityKey,
            hashIterations: hashIterations
        )
    }

    public enum MatchResult {
        case match
        case theyHaveOldVersion(localizedErrorDescription: String)
        case weHaveOldVersion(localizedErrorDescription: String)
        case noMatch(localizedErrorDescription: String)
    }

    public func matchesLogicalFingerprintsData(_ otherData: Data) -> MatchResult {
        owsAssertDebug(otherData.isEmpty.negated)

        let logicalFingerprints: FingerprintProtoLogicalFingerprints
        do {
            logicalFingerprints = try FingerprintProtoLogicalFingerprints.init(serializedData: otherData)
        } catch {
            owsFailDebug("fingerprint failure: \(error)")
            let description = OWSLocalizedString("PRIVACY_VERIFICATION_FAILURE_INVALID_QRCODE", comment: "alert body")
            return .noMatch(localizedErrorDescription: description)
        }

        if logicalFingerprints.version < self.scannableFingerprintVersion {
            Logger.warn("Verification failed. They're running an old version.")
            let description = OWSLocalizedString("PRIVACY_VERIFICATION_FAILED_WITH_OLD_REMOTE_VERSION", comment: "alert body")
            return .theyHaveOldVersion(localizedErrorDescription: description)
        }

        if logicalFingerprints.version > self.scannableFingerprintVersion {
            Logger.warn("Verification failed. We're running an old version.")
            let description = OWSLocalizedString("PRIVACY_VERIFICATION_FAILED_WITH_OLD_LOCAL_VERSION", comment: "alert body")
            return .weHaveOldVersion(localizedErrorDescription: description)
        }

        // Their local is *our* remote.
        let localFingerprint = logicalFingerprints.remoteFingerprint
        let remoteFingerprint = logicalFingerprints.localFingerprint
        if remoteFingerprint.identityData != Self.scannableData(from: self.theirFingerprintData) {
            Logger.warn("Verification failed. We have the wrong fingerprint for them")
            let descriptionFormat = OWSLocalizedString(
                "PRIVACY_VERIFICATION_FAILED_I_HAVE_WRONG_KEY_FOR_THEM",
                comment: "Alert body when verifying with {{contact name}}"
            )
            let description = String(format: descriptionFormat, self.theirName)
            return .noMatch(localizedErrorDescription: description)
        }
        if localFingerprint.identityData != Self.scannableData(from: self.myFingerprintData) {
            Logger.warn("Verification failed. They have the wrong fingerprint for us")
            let descriptionFormat = OWSLocalizedString(
                "PRIVACY_VERIFICATION_FAILED_THEY_HAVE_WRONG_KEY_FOR_ME",
                comment: "Alert body when verifying with {{contact name}}"
            )
            let description = String(format: descriptionFormat, self.theirName)
            return .noMatch(localizedErrorDescription: description)
        }

        Logger.warn("Verification Succeeded.")
        return .match
    }

    // MARK: - Text Representation

    private var textRepresentation: String {
        let myDisplayString = Self.stringForFingerprintData(myFingerprintData)
        let theirDisplayString = Self.stringForFingerprintData(theirFingerprintData)

        if theirDisplayString.compare(myDisplayString) == .orderedAscending {
            return theirDisplayString + myDisplayString
        } else {
            return myDisplayString + theirDisplayString
        }
    }

    private func generateDisplayableText() -> String {
        let input = self.textRepresentation

        var lines = [String]()

        let lineLength = (input as NSString).length / 3
        for i in 0..<3 {
            let line = input.substring(withRange: NSRange(location: i * lineLength, length: lineLength))
            var chunks = [String]()
            for j in 0..<((line as NSString).length / 5) {
                let nextChunk = line.substring(withRange: NSRange(location: j * 5, length: 5))
                chunks.append(nextChunk)
            }
            lines.append(String(chunks.joined(separator: " ")))
        }
        return String(lines.joined(separator: "\n"))
    }

    // MARK: - Image Representation

    private func generateImage() -> UIImage? {
        let remoteFingerprintBuilder = FingerprintProtoLogicalFingerprint.builder(
            identityData: Self.scannableData(from: self.theirFingerprintData)
        )
        let localFingerprintBuilder = FingerprintProtoLogicalFingerprint.builder(
            identityData: Self.scannableData(from: self.myFingerprintData)
        )
        let remoteFingerprint: FingerprintProtoLogicalFingerprint
        let localFingerprint: FingerprintProtoLogicalFingerprint
        do {
            remoteFingerprint = try remoteFingerprintBuilder.build()
            localFingerprint = try localFingerprintBuilder.build()
        } catch {
            owsFailDebug("could not build proto \(error)")
            return nil
        }

        let logicalFingerprintsBuilder = FingerprintProtoLogicalFingerprints.builder(
            version: self.scannableFingerprintVersion,
            localFingerprint: localFingerprint,
            remoteFingerprint: remoteFingerprint
        )

        let fingerprintData: Data
        do {
            // Build ByteMode QR (Latin-1 encodable data)
            fingerprintData = try logicalFingerprintsBuilder.buildSerializedData()
        } catch {
            owsFailDebug("could not serialize proto \(error)")
            return nil
        }

        Logger.debug("Building fingerprint")

        guard let filter = CIFilter(name: "CIQRCodeGenerator") else {
            Logger.error("Failed to create QR code filter")
            return nil
        }
        filter.setDefaults()
        filter.setValue(fingerprintData, forKey: "inputMessage")
        guard let ciImage = filter.outputImage else {
            Logger.error("Failed to create QR image from fingerprint")
            return nil
        }

        // UIImages backed by a CIImage won't render without antialiasing, so we convert the backign image to a CGImage,
        // which can be scaled crisply.
        let context = CIContext()
        guard let cgImage = context.createCGImage(ciImage, from: ciImage.extent) else {
            return nil
        }
        let qrImage = UIImage(cgImage: cgImage)
        return qrImage
    }

    // MARK: - Private helpers

    private static func stableData(myAci: Aci, theirAci: Aci) -> (my: Data, their: Data) {
        return (my: myAci.rawUUID.data, their: theirAci.rawUUID.data)
    }

    /**
     * An identifier for a mutable public key, belonging to an immutable identifier (stableId).
     *
     * This method is intended to be somewhat expensive to produce in order to be brute force adverse.
     *
     * @param stableAddressData
     *      Immutable global identifier e.g. Signal Identifier, an e164 formatted phone number encoded as UTF-8 data
     * @param publicKey
     *      The current public key for <stableAddress>
     * @return
     *      All-number textual representation
     */
    private static func dataForStableAddress(_ stableAddressData: Data, publicKey: Data, hashIterations: UInt32) -> Data {
        var hash = Constants.hashingVersion.bigEndianData.suffix(2)
        hash.append(publicKey)
        hash.append(stableAddressData)

        var digestData = Data(count: Int(CC_SHA512_DIGEST_LENGTH))

        for _ in 0..<hashIterations {
            hash.append(publicKey)
            if hash.count >= UInt32.max {
                owsFail("Oversize data")
            }

            digestData.withUnsafeMutableBytes({ mutableBufferPointer in
                let bufferPointer = mutableBufferPointer.bindMemory(to: UInt8.self)
                if let bufferAddress = bufferPointer.baseAddress {
                    hash.withUnsafeBytes { hashBytesPointer in
                        let hashPointer = hashBytesPointer.bindMemory(to: UInt8.self)
                        if let hashAddress = hashPointer.baseAddress {
                            CC_SHA512(hashAddress, CC_LONG(hash.count), bufferAddress)
                        }
                    }
                }
            })
            hash = digestData
        }

        return hash
    }

    private static func stringForFingerprintData(_ data: Data) -> String {
        return String(
            format: "%@%@%@%@%@%@",
            encodedChunkFromData(data, offset: 0),
            encodedChunkFromData(data, offset: 5),
            encodedChunkFromData(data, offset: 10),
            encodedChunkFromData(data, offset: 15),
            encodedChunkFromData(data, offset: 20),
            encodedChunkFromData(data, offset: 25)
        )
    }

    private static func encodedChunkFromData(_ data: Data, offset: Int) -> String {
        let fiveByteChunk = Data(data.dropFirst(offset).prefix(5))
        let chunk: Int = intFrom5Bytes(fiveByteChunk) % 100000
        return String(format: "%05d", chunk)
    }

    private static func intFrom5Bytes(_ data: Data) -> Int {
        return Int(data[0]) << 32
            + Int(data[1]) << 24
            + Int(data[2]) << 16
            + Int(data[3]) << 8
            + Int(data[4])
    }

    private static func scannableData(from data: Data) -> Data {
        return data.prefix(32)
    }

    private var scannableFingerprintVersion: UInt32 {
        return Constants.aciScannableFormatVersion
    }

    public enum Constants {
        static let hashingVersion: UInt32 = 0
        static let aciScannableFormatVersion: UInt32 = 2
        public static let defaultHashIterations: UInt32 = 5200
    }
}
